repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -82,21 +84,46 @@ class Session(interpreter: Interpreter, stateChangedCallback: SessionState => Un
   def execute(code: String): Int = {
     val statementId = newStatementId.getAndIncrement()
     _statements(statementId) = new Statement(statementId, StatementState.Waiting, null)
+
     Future {
-      _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+      _statements(statementId).state.compareAndSet(StatementState.Waiting, StatementState.Running)
 
-      _statements(statementId) =
-        new Statement(statementId, StatementState.Available, executeCode(statementId, code))
+      setJobGroup(statementId)
+      val executeResult = if (_statements(statementId).state.get() != StatementState.Cancelling) {
```

Comment:
There's a race condition here. Consider the following execution orders:
Interpreter execution thread:
Session.execute() [line 92](https://github.com/cloudera/livy/pull/230/files#diff-726504ec4098560b7f98fdec97b165e7R92) state is StatementState.Running, will execute line 93.

Cancelling thread:
Session.cancel() [line 109](https://github.com/cloudera/livy/pull/230/files#diff-726504ec4098560b7f98fdec97b165e7R109)
oldStmtState =  StatementState.Runnig
Session.cancel() calls `_sc.foreach(_.cancelJobGroup(statementId.toString))` [line 113-114](https://github.com/cloudera/livy/pull/230/files#diff-726504ec4098560b7f98fdec97b165e7R114)

Interpreter execution thread:
Session.execute() calls executeCode() [line 93](https://github.com/cloudera/livy/pull/230/files#diff-726504ec4098560b7f98fdec97b165e7R93)

Because cancelJobGroup() is called before executeCode() is executed, calls to Spark will be executed.